File: src/383_Ransom-Note/First.py
# Using list in python as hashmap will be simple
# 146 ms(< 14.38%), 14.5 MB(< 5.25%)

class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        hashmap = []
        for i in magazine:
            hashmap.append(i)
        for i in ransomNote:
            try:
                hashmap.remove(i)
            except ValueError:
                return False
        return True

File: src/383_Ransom-Note/First.c
// My First Solution... kind of a stupid program.
// 4 ms(< 88.2%), 6.6 MB(< 40.02%)

bool canConstruct(char * ransomNote, char * magazine){
    int alphabet[26] = {0};
    char* ptr = magazine;
    while(*ptr)
        alphabet[*(ptr++) - 'a']++;
    ptr = ransomNote;
    while(*ptr)
        if(--alphabet[*(ptr++) - 'a'] == -1)
            return false;
    return true;
}

// encountered bugs:
// runtime error: index -376483328 out of bounds for type 'int [26]' [solution.c]
// Cause: while(*ptr) not while(ptr)
//
// alphabet[*(ptr++) - 'a'] better than alphabet[(int)*(ptr++) - 95]
